### PR TITLE
[Bug][Spec] fix startup crash and reduce CUDA graph memory usage for speculative adaptive

### PR DESCRIPTION
**Pull request by Deleter-D**
## Motivation

Three separate problems show up when `--speculative-adaptive` is enabled:
1. **Startup crash.** `ModelRunner.max_decode_logits_rows()` sizes the shared logits buffer from the *static* `decode_num_tokens_per_req()`, but adaptive speculative decoding replays runners built for larger draft-token widths. The buffer is too small and the server fails to start. (mentioned in #30549)
2. **IMA on the first replay.** `DecodeCudaGraphRunner.capture_one_shape` looked up `on_after_cuda_graph_warmup` on `self.model_runner.attn_backend` instead of on the backend this capture actually ran against (`attn_backend` from `capture_prepare`). For the runner the `ModelRunner` owns these are the same object, but adaptive speculative decoding builds extra runners bound to their own freshly created backend while `model_runner.attn_backend` still points at the active one. Hooking the wrong backend leaves this capture's warmup-upgraded metadata (raw→full, FlashMLA scheduler meta) frozen into the graph, which produces an illegal memory access on the first replay.
3. **Redundant CUDA graph memory.** Each capture pass opened `graph_capture()` without a stream, so every pass captured on a fresh stream. The caching allocator partitions the graph pool's segments **by stream**, so each pass re-reserved the pool's MoE / DeepEP scratch instead of reusing what an earlier pass left inactive. With adaptive speculative decoding that is one pass per candidate step × three runners — tens of GB of segments that all read back `inactive`.

## Modifications
1. `max_decode_logits_rows()` now derives `num_tokens_per_req` from `max_speculative_num_draft_tokens`, so the shared logits buffer is sized for the widest adaptive graph. (Same fix as #30554.)
2. `decode_cuda_graph_runner.py:1226-1237` — resolve `post_warmup_hook` from the capture-local `attn_backend` rather than `model_runner.attn_backend`, with a comment recording why.
3. Added `get_or_create_global_graph_capture_stream()` in `runner_utils/pool.py` (a named stream lease in the `Resources` tier, alongside the existing process-wide graph memory pool) and used it as the `graph_capture(stream=...)` argument in both `DecodeCudaGraphRunner.capture` and `PrefillCudaGraphRunner.capture`. All capture passes therefore share one stream and reserve that scratch exactly once. Captures are strictly serial (`Scheduler.init_all_cuda_graphs` runs target and draft workers in turn; the adaptive controller builds one runtime state at a time), so a single shared stream is safe.

Also added a CPU unit test `test/registered/unit/model_executor/test_graph_capture_resources.py` covering both the pool and the new capture-stream lease.

## Accuracy Tests

Server launched following the cookbook:

```bash
sglang serve \
    --trust-remote-code \
    --model-path deepseek-ai/DeepSeek-V4-Flash \
    --tp 4 \
    --moe-runner-backend flashinfer_mxfp4 \
    --speculative-algorithm EAGLE \
    --speculative-num-steps 3 \
    --speculative-eagle-topk 1 \
    --speculative-num-draft-tokens 4 \
    --speculative-adaptive \
    --disable-flashinfer-autotune \
    --swa-full-tokens-ratio 0.1 \
    --host 0.0.0.0 \
    --port 30000
```

Evaluated on the three datasets shown in the cookbook:

|          | GPQA Diamond | AIME25 | GSM8k |
| -------- | ------------ | ------ | ----- |
| cookbook | 88.1         | 95     | 96.13 |
| before   | 88.13        | 98.33  | 96.89 |
| after    | 88.04          | 97.29  | 96.97 |

## Speed Tests and Profiling

No throughput change is expected; the win is memory. With the launch command above and the default adaptive speculative decoding parameters, CUDA graph memory drops from **14.4 GB → 10.13 GB**. With my own serving config and adaptive speculative decoding parameters, it drops from **44.3 GB → 16 GB**, which frees the corresponding capacity for the KV cache.

## Checklist

- [x] Format your code according to the [Format code with pre-commit](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [x] Add unit tests according to the [Run and add unit tests](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [ ] Update documentation according to [Write documentations](https://docs.sglang.io/developer_guide/contribution_guide.html#write-documentations).
- [x] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.io/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.io/developer_guide/contribution_guide.html#benchmark-the-speed).
- [x] Follow the SGLang code style [guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#code-style-guidance).

## Review and Merge Process

1. Ping Merge Oncalls to start the process. See the [PR Merge Process](https://github.com/sgl-project/sglang/blob/main/.github/MAINTAINER.md#pull-request-merge-process).
2. Get approvals from [CODEOWNERS](https://github.com/sgl-project/sglang/blob/main/.github/CODEOWNERS) and other reviewers.
3. Trigger CI tests with [comments](https://docs.sglang.io/developer_guide/contribution_guide.html#how-to-trigger-ci-tests) or contact authorized users to do so.
   - Common commands include `/tag-and-rerun-ci`, `/tag-run-ci-label`, `/rerun-failed-ci`
4. After green CI and required approvals, ask Merge Oncalls or people with Write permission to merge the PR.

---
### CI States

Latest PR Test (Base): :white_check_mark: [Run #32804934374](https://github.com/sgl-project/sglang/actions/runs/32804934374)
Latest PR Test (Extra): :x: [Run #32804934225](https://github.com/sgl-project/sglang/actions/runs/32804934225)
Latest PR Test (AMD ROCm 7.2): :x: [Run #32804934346](https://github.com/sgl-project/sglang/actions/runs/32804934346)

### CODE REVIEW

File: python/sglang/srt/model_executor/runner_utils/pool.py [hunk 1]
```diff
@@ -38,3 +46,22 @@ def get_or_create_global_graph_memory_pool(device_module: Any) -> Any:
     if resources.graph_memory_pool is None:
         resources.graph_memory_pool = device_module.graph_pool_handle()
     return resources.graph_memory_pool
+
+
+def get_or_create_global_graph_capture_stream() -> Any:
```
Comment (hnyls2002) [inline python/sglang/srt/model_executor/runner_utils/pool.py:87]: Too many comments. Did you review every line of the comments manually?

File: python/sglang/srt/model_executor/runner/decode_cuda_graph_runner.py [hunk 2]
```diff
@@ -1212,8 +1223,18 @@ def run_once():
                     variant_label,
                     dsa_variant,
                 )
+                # NOTE (Deleter-D)
```
Comment (hnyls2002) [inline python/sglang/srt/model_executor/runner/decode_cuda_graph_runner.py]: Only one line change; normally, our comments require only a few lines.

File: test/registered/unit/model_executor/test_graph_capture_resources.py [hunk 3]
```diff
@@ -0,0 +1,98 @@
+"""Unit tests for the process-wide CUDA graph capture resources (``pool.py``).
+
+The capture *stream* is shared for the same reason the memory pool is: the
+caching allocator partitions the graph pool's segments by stream, so a fresh
+stream per capture pass re-reserves that pool's MoE / DeepEP scratch instead of
+reusing what an earlier pass left inactive. With adaptive speculative decoding
+that is one pass per candidate step times three runners — tens of GB of
+segments that all read back ``inactive``.
+
+CPU-only: ``get_stream`` allocates a ``torch.cuda.Stream`` handle, which does
```
Comment (alphabetc1) [inline test/registered/unit/model_executor/test_graph_capture_resources.py]: It looks like it stilll relies on the CUDA runtime so can't run on CPU-only test?

https://github.com/sgl-project/sglang/actions/runs/32329277658/job/96306667802

Comment (Deleter-D) [reply test/registered/unit/model_executor/test_graph_capture_resources.py]: Thanks for catching this. `TestGraphCaptureStream` now patches torch.cuda.Stream with a fake — the assertions cover the lease logic in `pool.py`, not the stream object itself, so the semantics are unchanged. Verified locally on a machine with no visible device via 

`CUDA_VISIBLE_DEVICES=99 python test/registered/unit/model_executor/test_graph_capture_resources.py -v`

all 6 cases pass, and no CUDA context is initialized during the run.

**Review by alphabetc1 — APPROVED** [review]

**Review by hnyls2002 — APPROVED** [review]
